### PR TITLE
Add tracing to Truss Server, fixes BT-11854

### PR DESCRIPTION
**Pull request by marius-baseten**
## :rocket: What
Instrument Truss server with Otel tracing.

* Differentiating various sections such as data serialization, pre/post-processing and recording wall times.
* Special care has been taken to separate trace context (i.e. parent/child-relationships of nested traces) from our internal tracing instrumentation and optional user-space tracing (that might not be aware of needing to create "clean" contexts).
* For async streaming, the tracing context is tied to together with the deferred release of the semaphore.
* Structure the various predict return use cases into helper functions.

Side Quests:
* Make some unit-tests platform portable for correct local execution.
* Tentatively change the builder logic to always include the server requirements. The previous logic makes sense too, but is debatable... in this case it just makes it easier to get a consistent image without needing to update the base image externally. The whole point of image generation might need to be revisited on it's own. 

image

## :microscope: Testing
Added unitests.
Test with deployed models (see screenshot).

### CODE REVIEW

File: truss/templates/server/common/tracing.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+from opentelemetry import context, trace
+from shared import secrets_resolver
+
+logger = logging.getLogger(__name__)
+
+ATTR_NAME_DURATION = "duration_sec"
+OTEL_EXPORTER_OTLP_ENDPOINT = "OTEL_EXPORTER_OTLP_ENDPOINT"
+OTEL_TRACING_NDJSON_FILE = "OTEL_TRACING_NDJSON_FILE"
+HONEYCOMB_DATASET = "HONEYCOMB_DATASET"
+HONEYCOMB_API_KEY = "HONEYCOMB_API_KEY"
```
[the remark this answers is not in the document]
Comment (marius-baseten) [reply truss/templates/server/common/tracing.py:22]: Essentially honeycomb is just for debugging here. The "canonical" export path is via `OTEL_EXPORTER_OTLP_ENDPOINT` which is supposed to be provided as a pod env var.

[the remark this answers is not in the document]
Comment (squidarth) [reply truss/templates/server/common/tracing.py:29]: from a higher-level, who sets this `OTEL_EXPORTER_OTLP_ENDPOINT` endpoitn?

[the remark this answers is not in the document]
Comment (marius-baseten) [reply truss/templates/server/common/tracing.py:29]: We'll set it in the model pod (follow up PR): https://github.com/basetenlabs/baseten/pull/9166/files#diff-a71e7f44ad699dd30040d890b6e8c67355cdc4ec362c3f0a5eeb9ce65e2804d2

File: truss/contexts/image_builder/serving_image_builder.py [hunk 2]
```diff
@@ -459,10 +459,13 @@ def copy_into_build_dir(from_path: Path, path_in_build_dir: str):
...
         # requirements provided, we just pass an empty string,
         # as there's no need to install anything.
+        # TODO: above reasoning leads to inconsistencies. To get consistent images
+        #  tentatively add server requirements always. This whole point needs more
+        #  thought and potentially a re-design.
         user_provided_python_requirements = (
             base_server_requirements + spec.requirements_txt
             if spec.requirements
-            else ""
+            else base_server_requirements
```
[the remark this answers is not in the document]
Comment (marius-baseten) [reply truss/contexts/image_builder/serving_image_builder.py:468]: So should we rephrase the comment above?
I can somehow understand the original reasoning, but it's TBH a bit unprincipled *either way*...

File: truss/templates/server/common/tracing.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    enable_tracing_data = config.get("runtime", {}).get(
+        "enable_tracing_data", DEFAULT_ENABLE_TRACING_DATA
+    )
+
+    global _truss_tracer
+    if _truss_tracer:
+        return _truss_tracer
+
+    span_processors: List[sdk_trace.SpanProcessor] = []
+    if otlp_endpoint := os.getenv(OTEL_EXPORTER_OTLP_ENDPOINT):
```
[the remark this answers is not in the document]
Comment (marius-baseten) [reply truss/templates/server/common/tracing.py:74]: Honeycomb env vars can be set via truss config (they are for debugging / manual testing only, not as a production default.

The exporter endpoints will be set as env var in the model pods.

**Pull request merged**